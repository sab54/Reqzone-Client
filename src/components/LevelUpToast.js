/**
 * LevelUpToast.js
 *
 * This component displays a transient “level up” toast that fades in/out using an external Animated.Value.
 *
 * What This Component Does:
 * - **Animated Visibility**: Uses the provided `animatedValue` (0–1) to drive the toast’s `opacity`.
 * - **Theming**: Pulls `backgroundColor` from `theme.card`, text color from `theme.primary`,
 *   and shadow color from `theme.shadow` (fallback `#000`).
 * - **Status Message**: Shows a celebratory message with the new `level`.
 *
 * Component Structure:
 * 1. **Animated.View**: Positioned absolutely near the bottom-center with rounded corners and shadow/elevation.
 * 2. **Text**: Bold-ish Poppins font with primary-colored text describing the level-up.
 *
 * Props:
 * - `animatedValue: Animated.AnimatedInterpolation | Animated.Value`
 *    - Controls the opacity of the toast. Expected range: 0 (hidden) to 1 (fully visible).
 * - `level: number | string`
 *    - The target level to display in the message.
 * - `theme: { card: string, primary: string, shadow?: string }`
 *    - `card`: Background color of the toast container.
 *    - `primary`: Color for the text.
 *    - `shadow` (optional): Shadow color for iOS; defaults to `#000`.
 *
 * Behavior:
 * - Visual only: no timers or gesture handling here. Parent component is responsible for animating `animatedValue`
 *   (e.g., via `Animated.timing`) and for mounting/unmounting lifecycle.
 * - Layout is absolute with fixed insets to sit above other UI near the bottom.
 *
 * Notes:
 * - Requires `PoppinsBold` (or alias) to be loaded in the app for correct font rendering.
 * - Shadow props (iOS) and `elevation` (Android) are applied for subtle depth.
 * - Keep performance in mind: drive the opacity with `useNativeDriver: true` in parent animations where possible.
 *
 * Author: Sunidhi Abhange
 */

import React from 'react';
import { Animated, Text, StyleSheet } from 'react-native';

const LevelUpToast = ({ animatedValue, level, theme }) => (
    <Animated.View
        style={[
            styles.toast,
            {
                opacity: animatedValue,
                backgroundColor: theme.card,
                shadowColor: theme.shadow || '#000',
            },
        ]}
    >
        <Text style={[styles.text, { color: theme.primary }]}>
            🎉 Level Up! You're now Level {level}!
        </Text>
    </Animated.View>
);

const styles = StyleSheet.create({
    toast: {
        position: 'absolute',
        bottom: 100,
        left: 20,
        right: 20,
        padding: 15,
        borderRadius: 10,
        alignItems: 'center',
        elevation: 6,
        shadowOffset: { width: 0, height: 2 },
        shadowOpacity: 0.15,
        shadowRadius: 4,
    },
    text: {
        fontSize: 15,
        fontFamily: 'PoppinsBold',
    },
});

export default LevelUpToast;
