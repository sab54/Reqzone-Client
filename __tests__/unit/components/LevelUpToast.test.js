/**
 * LevelUpToast.test.js
 *
 * What This Test File Covers:
 *
 * 1. Rendering with props
 *    - Ensures the toast renders the correct message for a given `level`.
 *
 * 2. Theme application
 *    - Confirms that `theme.card` is applied as background and `theme.primary` is applied to text color.
 *
 * 3. Animated value usage
 *    - Asserts the `opacity` style reflects the numeric value of the provided Animated.Value.
 *      (Accepts either a raw number or an Animated node, depending on the RN test env/mocks.)
 *
 * 4. Snapshot
 *    - Captures stable output for regression testing.
 *
 * Notes:
 * - No component changes required; this test adapts to environments where Animated flattens to a number.
 */

import React from 'react';
import { Animated } from 'react-native';
import { render } from '@testing-library/react-native';
import LevelUpToast from 'src/components/LevelUpToast';

const theme = {
  card: '#ffffff',
  primary: '#123456',
  shadow: '#654321',
};

describe('LevelUpToast', () => {
  it('renders the correct message for the given level', () => {
    const { getByText } = render(
      <LevelUpToast animatedValue={new Animated.Value(1)} level={7} theme={theme} />
    );
    expect(getByText("ðŸŽ‰ Level Up! You're now Level 7!")).toBeTruthy();
  });

  it('applies theme colors to container and text', () => {
    const animatedValue = new Animated.Value(0.8);
    const { getByText, toJSON } = render(
      <LevelUpToast animatedValue={animatedValue} level={5} theme={theme} />
    );

    // Text color from theme.primary
    const text = getByText("ðŸŽ‰ Level Up! You're now Level 5!");
    expect(text.props.style).toEqual(
      expect.arrayContaining([expect.objectContaining({ color: theme.primary })])
    );

    // Container background from theme.card
    const tree = toJSON();
    const containerStyle = tree.props.style;
    const flattened = Array.isArray(containerStyle)
      ? Object.assign({}, ...containerStyle)
      : containerStyle;
    expect(flattened.backgroundColor).toBe(theme.card);
  });

  it('uses the provided animated value for opacity (accepts number or Animated node)', () => {
    const animatedValue = new Animated.Value(0.5);
    const { toJSON } = render(
      <LevelUpToast animatedValue={animatedValue} level={10} theme={theme} />
    );

    const root = toJSON();
    const containerStyle = root.props.style;
    const flattened = Array.isArray(containerStyle)
      ? Object.assign({}, ...containerStyle)
      : containerStyle;

    // Resolve numeric value from style.opacity; RN test env may give number or Animated node
    const resolve = (v) => (typeof v === 'number' ? v : v?.__getValue?.());
    const resolvedOpacity = resolve(flattened.opacity);

    expect(typeof resolvedOpacity).toBe('number');
    expect(resolvedOpacity).toBeCloseTo(0.5, 5);

    // Sanity check: source value also holds 0.5
    expect(animatedValue.__getValue()).toBeCloseTo(0.5, 5);
  });

  it('matches snapshot', () => {
    const tree = render(
      <LevelUpToast animatedValue={new Animated.Value(1)} level={3} theme={theme} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
