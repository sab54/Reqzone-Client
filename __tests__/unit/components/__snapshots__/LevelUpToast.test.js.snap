// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LevelUpToast matches snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "borderRadius": 10,
      "bottom": 100,
      "elevation": 6,
      "left": 20,
      "opacity": 1,
      "padding": 15,
      "position": "absolute",
      "right": 20,
      "shadowColor": "#654321",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        {
          "fontFamily": "PoppinsBold",
          "fontSize": 15,
        },
        {
          "color": "#123456",
        },
      ]
    }
  >
    ðŸŽ‰ Level Up! You're now Level 
    3
    !
  </Text>
</View>
`;
